challenge-manager/challenge-tree/tree.go

```diff
@@ -56,14 +59,24 @@ func New(
 		histChecker:           histChecker,
 		validatorName:         validatorName,
 		// The total number of challenge levels include block challenges, small step challenges, and N big step challenges.
-		totalChallengeLevels:   numBigStepLevels + 2,
-		honestRootEdgesByLevel: threadsafe.NewMap[protocol.ChallengeLevel, *threadsafe.Slice[protocol.ReadOnlyEdge]](),
+		totalChallengeLevels:          numBigStepLevels + 2,
+		honestRootEdgesByLevel:        threadsafe.NewMap[protocol.ChallengeLevel, *threadsafe.Slice[protocol.ReadOnlyEdge]](),
+		agreeWithStartCommitmentCache: cache,
 	}
 }
 
 // AddEdge to the honest challenge tree. Only honest edges are tracked, but we also keep track
 // of rival ids in a mutual ids mapping internally for extra book-keeping.
 func (ht *HonestChallengeTree) AddEdge(ctx context.Context, eg protocol.SpecEdge) (protocol.Agreement, error) {
+	// Check if we already agree with the start commitment of this edge.
+	startHeight, startCommit := eg.StartCommitment()
+	if _, ok := ht.agreeWithStartCommitmentCache.Get(startCommit); ok {
```

Comment:
A commit is more than the merkle. It also encompasses the height. Our cache key should include both